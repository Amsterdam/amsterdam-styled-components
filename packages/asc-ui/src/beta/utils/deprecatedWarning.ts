/**
 * A general deprecated console warning avoided by lint
 */

export default (message: string) => {
  // eslint-disable-next-line
  console.warn(message)
}
