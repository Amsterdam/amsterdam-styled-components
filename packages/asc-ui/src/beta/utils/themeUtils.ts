import { css } from 'styled-components'
import type { Theme } from '../types'
import { fromProps } from './fromProps'
import { ascBetaTheme } from '..'

export type ThemeFn<T> = ({ theme }: { theme: Theme.ThemeInterface }) => T

/**
 * Curry function to provide the theme as first parameter, followed up with other parameters
 * @param cb
 *
 * @example
 * const myThemeHelperFunction = withTheme((theme, ...otherParams) => {
 *   ...
 * })
 *
 * css`
 *  ${myThemeHelperFunction(param1, param2)}
 * `
 */
export const withTheme =
  <T extends any[], U = any>(
    cb: (theme: Theme.ThemeInterface, ...params: T) => U,
  ) =>
  (...params: T) =>
  ({ theme }: { theme: Theme.ThemeInterface }): U =>
    cb(theme, ...params)

type ThemeColorParameters = [Theme.ColorType?, string?, string?]

/**
 * A shortcut to the `fromProps` that will get a value out of the props.theme object
 */
export const getValueFromTheme = withTheme<[string, ((value: any) => void)?]>(
  (theme, identifier, callback?: (value: any) => void) =>
    fromProps(identifier, callback)(theme),
)

export const themeColor = withTheme<ThemeColorParameters, string>(
  (theme, colorType, colorSubtype = 'main', override) => {
    if (override) {
      return override
    }

    if (colorType) {
      return getValueFromTheme(`colors.${[colorType]}.${[colorSubtype]}`)({
        theme,
      })
    }

    return getValueFromTheme('colors.tint.level1')({ theme })
  },
)

type BreakpointsType = [Theme.TypeBreakpoint, keyof Theme.BreakpointsInterface]

export const breakpoint = withTheme<BreakpointsType>((theme, type, variant) => {
  const breakpointFunc: Theme.GetBreakpointFunc = getValueFromTheme(
    `breakpoints.${[variant]}`,
  )({
    theme,
  })
  return breakpointFunc && breakpointFunc(type)
})

/**
 * When this style is applied on an element it will be hidden but still readable by screen readers.
 *
 * For example:
 * ```ts
 * const OnlyVisibleForScreenReaders = styled.button`${srOnlyStyle}`
 * ```
 */
export const srOnlyStyle = css`
  border-width: 0;
  clip: rect(0, 0, 0, 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
`

/**
 * Fills the elements in an SVG with a color, useful for styling icons.
 *
 * For example, using a theme color:
 * ```jsx
 * const PrimaryCarIcon = styled(CarIcon)`
 *   ${svgFill(themeColor('primary'))}
 * `
 * ```
 *
 * Or by setting the color directly using any color valid in CSS, such as hexadecimal, rgba() and hsla().
 *
 * ```jsx
 * const RedCarIcon = styled(CarIcon)`
 *   ${svgFill('#FF0000')}
 * `
 * ```
 */
export const svgFill = withTheme((theme, color: string | ThemeFn<string>) => {
  const fill = typeof color === 'function' ? color({ theme }) : color

  return css`
    & svg {
      circle,
      rect,
      polygon,
      path {
        fill: ${fill};
      }
    }
  `
})

export const calculateFluidStyle = (
  minSizePx: number,
  maxSizePx: number,
  minScreenWidthPx = 320,
  maxScreenWidthPx = 1920,
) => {
  const defaultBaseSize = 16
  const minSize = minSizePx / defaultBaseSize
  const maxSize = maxSizePx / defaultBaseSize
  const minScreenWidth = minScreenWidthPx / defaultBaseSize
  const maxScreenWidth = maxScreenWidthPx / defaultBaseSize

  return `clamp(
    ${minSize}rem,
    ${minSize}rem + ${maxSize - minSize} * 
    (100vw - ${minScreenWidth}rem) / ${maxScreenWidth - minScreenWidth},
    ${maxSize}rem
  )`
}

export interface FluidTypoStyle {
  small?: boolean
  intro?: boolean
  styleAs?: any
  as?: any
}

export const fluidTypoStyle = css`
  ${({ small, intro, styleAs, as }: FluidTypoStyle) => {
    const tag =
      (Object.keys(ascBetaTheme.typography).includes(styleAs) && styleAs) ||
      (Object.keys(ascBetaTheme.typography).includes(as) && as) ||
      'p'

    const prop =
      (small &&
        Object.keys(ascBetaTheme.typography[tag]).includes('small') &&
        'small') ||
      (intro &&
        Object.keys(ascBetaTheme.typography[tag]).includes('intro') &&
        'intro') ||
      'default'

    return css`
      font-size: ${calculateFluidStyle(
        ascBetaTheme.typography[tag][prop].minFontSize,
        ascBetaTheme.typography[tag][prop].maxFontSize,
      )};
      line-height: ${calculateFluidStyle(
        ascBetaTheme.typography[tag][prop].minLineHeight,
        ascBetaTheme.typography[tag][prop].maxLineHeight,
      )};
    `
  }}
`

export interface TypoStyle {
  gutterBottom?: {
    small: number
    large: number
  }
  strong?: boolean
  darkBackground?: boolean
}

export const typographyStyle = css<TypoStyle>`
  margin-top: 0;
  margin-bottom: ${({ gutterBottom }) =>
    `${
      typeof gutterBottom?.large === 'number'
        ? gutterBottom.large
        : gutterBottom || 0
    }px`};

  ${({ gutterBottom }) =>
    typeof gutterBottom?.small === 'number' &&
    css`
      @media screen and ${breakpoint('max-width', 'laptop')} {
        margin-bottom: ${gutterBottom.small}px;
      }
    `}

  ${({ strong }) =>
    strong &&
    css`
      font-weight: 700;
    `}

  ${({ darkBackground }) =>
    darkBackground &&
    css`
      color: white;
      ${svgFill('white')}
    `}
`

type ThemeSpacingParameters = [
  Theme.Spacing,
  Theme.Spacing?,
  Theme.Spacing?,
  Theme.Spacing?,
]

/**
 * Retrieve an amount of pixels by passing a factor that will multiply it by the theme's spacing unit
 *
 * @example If the theme's spacing unit is 4px:
 * css`
 *  padding: ${themeSpacing(1, 2, 1, 4)} // padding: 4px 8px 4px 16px;
 *  margin-bottom: ${themeSpacing(3)} // margin-bottom: 12px;
 * `
 */
export const themeSpacing = withTheme<ThemeSpacingParameters>(
  (theme, ...factors) => {
    const spacing: Theme.Spacing = getValueFromTheme('spacing')({ theme })
    return factors
      .map((factor) => factor && `${factor * spacing}px`)
      .join(' ')
      .trim()
  },
)
