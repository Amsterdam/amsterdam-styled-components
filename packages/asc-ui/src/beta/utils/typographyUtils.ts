import { css } from 'styled-components'
import { svgFill, breakpoint } from '.'

export const calculateFluidStyle = (
  minSizePx: number,
  maxSizePx: number,
  minScreenWidthPx = 320,
  maxScreenWidthPx = 1920,
) => {
  const defaultBaseSize = 16
  const minSize = minSizePx / defaultBaseSize
  const maxSize = maxSizePx / defaultBaseSize
  const minScreenWidth = minScreenWidthPx / defaultBaseSize
  const maxScreenWidth = maxScreenWidthPx / defaultBaseSize

  return `clamp(
    ${minSize}rem,
    ${minSize}rem + ${maxSize - minSize} * 
    (100vw - ${minScreenWidth}rem) / ${maxScreenWidth - minScreenWidth},
    ${maxSize}rem
  )`
}

export const fluidTypoConfig = {
  h1: {
    default: {
      minFontSize: 32,
      maxFontSize: 80,
      minLineHeight: 40,
      maxLineHeight: 88,
    },
    small: {
      minFontSize: 30,
      maxFontSize: 60,
      minLineHeight: 34,
      maxLineHeight: 64,
    },
  },
  h2: {
    default: {
      minFontSize: 28,
      maxFontSize: 56,
      minLineHeight: 32,
      maxLineHeight: 64,
    },
  },
  h3: {
    default: {
      minFontSize: 24,
      maxFontSize: 40,
      minLineHeight: 32,
      maxLineHeight: 48,
    },
  },
  h4: {
    default: {
      minFontSize: 22,
      maxFontSize: 32,
      minLineHeight: 32,
      maxLineHeight: 40,
    },
  },
  h5: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 26,
      maxLineHeight: 32,
    },
  },
  h6: {
    default: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 22,
      maxLineHeight: 24,
    },
  },
  p: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    intro: {
      minFontSize: 18,
      maxFontSize: 30,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    small: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
  a: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 26,
      maxLineHeight: 32,
    },
  },
  button: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    small: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 22,
      maxLineHeight: 24,
    },
  },
  blockquote: {
    default: {
      minFontSize: 24,
      maxFontSize: 40,
      minLineHeight: 32,
      maxLineHeight: 48,
    },
  },
  ul: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    small: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
  ol: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    small: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
  dl: {
    default: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
  table: {
    default: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
}

export interface FluidTypoStyle {
  small?: boolean
  intro?: boolean
  styleAs?: any
  as?: any
}

export const fluidTypoStyle = css`
  ${({ small, intro, styleAs, as }: FluidTypoStyle) => {
    const tag =
      (Object.keys(fluidTypoConfig).includes(styleAs) && styleAs) ||
      (Object.keys(fluidTypoConfig).includes(as) && as) ||
      'p'

    const prop =
      (small &&
        Object.keys(fluidTypoConfig[tag]).includes('small') &&
        'small') ||
      (intro &&
        Object.keys(fluidTypoConfig[tag]).includes('intro') &&
        'intro') ||
      'default'

    return css`
      font-size: ${calculateFluidStyle(
        fluidTypoConfig[tag][prop].minFontSize,
        fluidTypoConfig[tag][prop].maxFontSize,
      )};
      line-height: ${calculateFluidStyle(
        fluidTypoConfig[tag][prop].minLineHeight,
        fluidTypoConfig[tag][prop].maxLineHeight,
      )};
    `
  }}
`

export interface TypoStyle {
  gutterBottom?: {
    small: number
    large: number
  }
  strong?: boolean
  darkBackground?: boolean
}

export const typographyStyle = css<TypoStyle>`
  margin-top: 0;
  margin-bottom: ${({ gutterBottom }) =>
    `${
      typeof gutterBottom?.large === 'number'
        ? gutterBottom.large
        : gutterBottom || 0
    }px`};

  ${({ gutterBottom }) =>
    typeof gutterBottom?.small === 'number' &&
    css`
      @media screen and ${breakpoint('max-width', 'laptop')} {
        margin-bottom: ${gutterBottom.small}px;
      }
    `}

  ${({ strong }) =>
    strong &&
    css`
      font-weight: 700;
    `}

  ${({ darkBackground }) =>
    darkBackground &&
    css`
      color: white;
      ${svgFill('white')}
    `}
`
