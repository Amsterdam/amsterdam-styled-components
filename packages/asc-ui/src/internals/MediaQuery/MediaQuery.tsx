import React from 'react'
import styled from '@datapunt/asc-core'

export interface IBreakpointParams {
  min?: string
  max?: string
  children: any
}

export const hideAt = (breakpoints: any) => {
  const rulesets = Object.entries(breakpoints).reduce(
    (acc, [constraint, width]: any): any => [
      ...acc,
      {
        constraint,
        width,
        rules: `display: none;`,
      },
    ],
    [],
  )
  const css = rulesets.reduce(
    (cssString: any, { constraint, width, rules }: any) =>
      `${cssString} @media screen and (${constraint}-width: ${width}) { ${rules} }`,
    '',
  )
  return css
}

export const MediaQuery = ({ min, max, children }: IBreakpointParams) => {
  const css = hideAt({ min, max })
  const Component = styled('div')`
    ${css}
  `
  return <Component>{children}</Component>
}

export default MediaQuery
