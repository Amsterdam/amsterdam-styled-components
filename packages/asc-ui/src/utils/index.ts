export { default as fromTheme, valueFromObject } from './fromTheme'
export {
  color,
  getColorCode,
  getTypographyFromTheme,
  focusStyleOutline,
  srOnlyStyle,
  breakpoint,
  svgFill,
  showHide,
  ShowHideTypes,
  customCss,
  CustomCssPropsType,
  CustomCssPropsInterface,
} from './themeUtils'
