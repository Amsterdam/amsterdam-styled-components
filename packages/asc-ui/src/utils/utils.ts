import { AbstractNode, IconDefinition } from '@datapunt/asc-icons/lib/types'
import * as React from 'react'

export function log(message: string) {
  if (!(process && process.env && process.env.NODE_ENV === 'production')) {
    console.error(`[icons-react]: ${message}.`)
  }
}

export function isIconDefinition(target: any): target is IconDefinition {
  return (
    typeof target === 'object' &&
    typeof target.name === 'string' &&
    typeof target.icon === 'object'
  )
}

export function normalizeAttrs(attrs: Attrs = {}): Attrs {
  return Object.keys(attrs).reduce((acc: Attrs, key) => {
    const val = attrs[key]
    switch (key) {
      case 'class':
        acc.className = val
        delete acc.class
        break
      default:
        acc[key] = val
    }
    return acc
  }, {})
}

export interface Attrs {
  [key: string]: string
}

export class MiniMap<V> {
  get size() {
    return Object.keys(this.collection).length
  }

  private collection: { [key: string]: V } = {}

  clear(): void {
    this.collection = {}
  }

  delete(key: string): boolean {
    return delete this.collection[key]
  }

  get(key: string): V | undefined {
    return this.collection[key]
  }

  has(key: string): boolean {
    return Boolean(this.collection[key])
  }

  set(key: string, value: V): this {
    this.collection[key] = value
    return this
  }
}

export function generate(
  node: AbstractNode,
  key: string,
  rootProps?: { [key: string]: any } | false,
): any {
  if (!rootProps) {
    return React.createElement(
      node.tag,
      { key, ...normalizeAttrs(node.attrs) },
      (node.children || []).map((child, index) =>
        generate(child, `${key}-${node.tag}-${index}`),
      ),
    )
  }
  return React.createElement(
    node.tag,
    {
      key,
      ...normalizeAttrs(node.attrs),
      ...rootProps,
    },
    (node.children || []).map((child, index) =>
      generate(child, `${key}-${node.tag}-${index}`),
    ),
  )
}
