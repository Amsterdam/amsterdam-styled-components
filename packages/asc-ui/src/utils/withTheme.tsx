import * as React from 'react'
import { shallow, render, mount } from 'enzyme'
import { ThemeProvider } from '../index'

export const renderWithTheme = (component: JSX.Element) =>
  render(<ThemeProvider>{component}</ThemeProvider>)

export const shallowWithTheme = (component: JSX.Element) =>
  shallow(<ThemeProvider>{component}</ThemeProvider>)

export const mountWithTheme = (component: JSX.Element) =>
  mount(<ThemeProvider>{component}</ThemeProvider>)
