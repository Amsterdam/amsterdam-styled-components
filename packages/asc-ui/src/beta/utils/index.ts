import deprecatedWarning from './deprecatedWarning'

export {
  themeColor,
  srOnlyStyle,
  breakpoint,
  svgFill,
  fluidTypoStyle,
  typographyStyle,
  showAboveBackDrop,
  getValueFromTheme,
  themeSpacing,
} from './themeUtils'
export { valueFromObject } from './valueFromObject'
export { fromProps } from './fromProps'

export { deprecatedWarning }
