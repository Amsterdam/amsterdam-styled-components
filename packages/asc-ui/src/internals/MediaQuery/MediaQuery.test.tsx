import React from 'react'
import MediaQuery, { hideAt } from './MediaQuery'
import { renderWithTheme } from '../../utils/withTheme'
import 'jest-styled-components'

describe('hideAt', () => {
  it('should return the correct style', () => {
    const css = hideAt({ min: '600px', max: '1200px' })
    expect(css).toBe(
      ' @media screen and (min-width: 600px) { display: none; } @media screen and (max-width: 1200px) { display: none; }',
    )
  })
})

describe('MediaQuery', () => {
  it.only('should create a button with media queries', () => {
    const component = renderWithTheme(
      <MediaQuery min="300px" max="900px">
        <button type="button">Test</button>
      </MediaQuery>,
    )
    expect(component).toMatchSnapshot()
  })
})
