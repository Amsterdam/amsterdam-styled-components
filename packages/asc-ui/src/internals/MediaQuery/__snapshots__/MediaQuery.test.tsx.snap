// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreakpointHOC should create a button with media queries 1`] = `
@media screen and (min-width:300px) {
  .c0 {
    display: none;
  }
}

@media screen and (max-width:900px) {
  .c0 {
    display: none;
  }
}

<div
  class="c0"
>
  <button
    type="button"
  >
    Test
  </button>
</div>
`;
